In [1]:
124341342**100

289437657159816160291931685301549134875715811806077725714830003776267785682274226323106051805486547024011325586702136571893734854579421745283047582729662329924773144958469051035430412291681389322950721231947002760531507991436975887512187526463391542865790918100245769074256225678385904247462307633129914169552670311238320130470112548798834847273871776502142578108656766613701809255132280020635524832458179979756270155824553156921817340238277741765822917998529364868416518364960675510751560541881622324553639568111328253146464861869591210746146451795947783018036749309527129315898351397252653774184946360182076167244286041082064399288472572185919972722591537843933722214441014345245468824982035957246466149084711932010191476337942460107449567826892613608562866706174146192076748912903100401652373978379112677376

# Exercise: working with Jupyter Notebook
1. Use the '+' botton to add new cell under the current one
2. type the following code and run it by using <kbd>Shift</kbd> + <kbd>Enter</kbd>:

        import pandas as pd
        pandas.read_csv('Data/Abalone.csv')
        
3. Getting used to Keyboard shortcuts:
        
     - <kbd>Esc</kbd>
     - <kbd>b</kbd>
     - <kbd>m</kbd>
     - <kbd>Enter</kbd>
     - <kbd>Ctrl</kbd> + <kbd>Shift</kbd>
     - <kbd>a</kbd>
     - <kbd>Enter</kbd>
     - <kbd>10**0.005</kbd>
     - <kbd>Shift</kbd> + <kbd>Enter</kbd>
     - <kbd>Esc</kbd>
     - <kbd>h</kbd>

In [7]:
import pandas as pd
pd.read_csv('Data/Abalone.csv')

Unnamed: 0 sex  length  diameter  height  whole weight  shucked weight  \
0              0   M   0.455     0.365   0.095        0.5140          0.2245   
1              1   M   0.350     0.265   0.090        0.2255          0.0995   
2              2   F   0.530     0.420   0.135        0.6770          0.2565   
3              3   M   0.440     0.365   0.125        0.5160          0.2155   
4              4   I   0.330     0.255   0.080        0.2050          0.0895   
5              5   I   0.425     0.300   0.095        0.3515          0.1410   
6              6   F   0.530     0.415   0.150        0.7775          0.2370   
7              7   F   0.545     0.425   0.125        0.7680          0.2940   
8              8   M   0.475     0.370   0.125        0.5095          0.2165   
9              9   F   0.550     0.440   0.150        0.8945          0.3145   
10            10   F   0.525     0.380   0.140        0.6065          0.1940   
11            11   M   0.430     0.350   0.110        0.4060          0.1675   
12            12   M   0.490     0.380   0.135        0.5415          0.2175   
13            13   F   0.535     0.405   0.145        0.6845          0.2725   
14            14   F   0.470     0.355   0.100        0.4755          0.1675   
15            15   M   0.500     0.400   0.130        0.6645          0.2580   
16            16   I   0.355     0.280   0.085        0.2905          0.0950   
17            17   F   0.440     0.340   0.100        0.4510          0.1880   
18            18   M   0.365     0.295   0.080        0.2555          0.0970   
19            19   M   0.450     0.320   0.100        0.3810          0.1705   
20            20   M   0.355     0.280   0.095        0.2455          0.0955   
21            21   I   0.380     0.275   0.100        0.2255          0.0800   
22            22   F   0.565     0.440   0.155        0.9395          0.4275   
23            23   F   0.550     0.415   0.135        0.7635          0.3180   
24            24   F   0.615     0.480   0.165        1.1615          0.5130   
25            25   F   0.560     0.440   0.140        0.9285          0.3825   
26            26   F   0.580     0.450   0.185        0.9955          0.3945   
27            27   M   0.590     0.445   0.140        0.9310          0.3560   
28            28   M   0.605     0.475   0.180        0.9365          0.3940   
29            29   M   0.575     0.425   0.140        0.8635          0.3930   
...          ...  ..     ...       ...     ...           ...             ...   
4147        4147   M   0.695     0.550   0.195        1.6645          0.7270   
4148        4148   M   0.770     0.605   0.175        2.0505          0.8005   
4149        4149   I   0.280     0.215   0.070        0.1240          0.0630   
4150        4150   I   0.330     0.230   0.080        0.1400          0.0565   
4151        4151   I   0.350     0.250   0.075        0.1695          0.0835   
4152        4152   I   0.370     0.280   0.090        0.2180          0.0995   
4153        4153   I   0.430     0.315   0.115        0.3840          0.1885   
4154        4154   I   0.435     0.330   0.095        0.3930          0.2190   
4155        4155   I   0.440     0.350   0.110        0.3805          0.1575   
4156        4156   M   0.475     0.370   0.110        0.4895          0.2185   
4157        4157   M   0.475     0.360   0.140        0.5135          0.2410   
4158        4158   I   0.480     0.355   0.110        0.4495          0.2010   
4159        4159   F   0.560     0.440   0.135        0.8025          0.3500   
4160        4160   F   0.585     0.475   0.165        1.0530          0.4580   
4161        4161   F   0.585     0.455   0.170        0.9945          0.4255   
4162        4162   M   0.385     0.255   0.100        0.3175          0.1370   
4163        4163   I   0.390     0.310   0.085        0.3440          0.1810   
4164        4164   I   0.390     0.290   0.100        0.2845          0.1255   
4165  

###### Trying some keyboard shortcuts

In [10]:
10**0.005

1.0115794542598986

In [11]:
print?

In [13]:
pd.read_csv?

In [14]:
open('Data/abalone.csv')

<_io.TextIOWrapper name='Data/abalone.csv' mode='r' encoding='cp1252'>

## Basic Python types

In [15]:
# integers: +ve or -ve whole numbers:

10086786

10086786

In [16]:
# floats: numbers with decimal points:

3.14

3.14

In [17]:
'C:\Users\ABD01F'

SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in position 2-3: truncated \UXXXXXXXX escape (<ipython-input-17-62f75e52288d>, line 1)

In [18]:
r'C:\Users\ABD01F'

'C:\\Users\\ABD01F'

In [19]:
print(r'C:\Users\ABD01F')

C:\Users\ABD01F


In [33]:
def great(name):
    print('Good morning ' + name + '!')

In [34]:
great('Muhammad')

Good morning Muhammad!


In [37]:
def imperial_to_metric(feet, inches): # positional arguments
    total_inches = feet * 12 + inches
    cms = total_inches * 2.54
    return(cms / 100)

In [44]:
height = imperial_to_metric(5, 6)

In [45]:
height

1.6764000000000001

In [46]:
def imperial_to_metric(feet=0, inches=0): # keyword arguments
    total_inches = feet * 12 + inches
    cms = total_inches * 2.54
    return(cms / 100)

In [47]:
imperial_to_metric(0, 10)

0.254

In [48]:
imperial_to_metric(inches=10)

0.254

In [77]:
def F_to_C(degrees_F, Kelvin_offset=0):
    degrees_C = (degrees_F-32) * (5/9)
    degrees_K = degrees_C + Kelvin_offset
    return degrees_C + degrees_K

In [78]:
F_to_C(-10, 273.15)

226.4833333333333

In [79]:
F_to_C(0, 273.15)

237.59444444444443

In [80]:
F_to_C(100, 273.15)

348.7055555555555

In [81]:
import numpy as np

temperatures = np.arange(-100, 100)

In [82]:
temperatures

array([-100,  -99,  -98,  -97,  -96,  -95,  -94,  -93,  -92,  -91,  -90,
        -89,  -88,  -87,  -86,  -85,  -84,  -83,  -82,  -81,  -80,  -79,
        -78,  -77,  -76,  -75,  -74,  -73,  -72,  -71,  -70,  -69,  -68,
        -67,  -66,  -65,  -64,  -63,  -62,  -61,  -60,  -59,  -58,  -57,
        -56,  -55,  -54,  -53,  -52,  -51,  -50,  -49,  -48,  -47,  -46,
        -45,  -44,  -43,  -42,  -41,  -40,  -39,  -38,  -37,  -36,  -35,
        -34,  -33,  -32,  -31,  -30,  -29,  -28,  -27,  -26,  -25,  -24,
        -23,  -22,  -21,  -20,  -19,  -18,  -17,  -16,  -15,  -14,  -13,
        -12,  -11,  -10,   -9,   -8,   -7,   -6,   -5,   -4,   -3,   -2,
         -1,    0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
         10,   11,   12,   13,   14,   15,   16,   17,   18,   19,   20,
         21,   22,   23,   24,   25,   26,   27,   28,   29,   30,   31,
         32,   33,   34,   35,   36,   37,   38,   39,   40,   41,   42,
         43,   44,   45,   46,   47,   48,   49,   

In [83]:
F_to_C(temperatures)

array([-146.66666667, -145.55555556, -144.44444444, -143.33333333,
       -142.22222222, -141.11111111, -140.        , -138.88888889,
       -137.77777778, -136.66666667, -135.55555556, -134.44444444,
       -133.33333333, -132.22222222, -131.11111111, -130.        ,
       -128.88888889, -127.77777778, -126.66666667, -125.55555556,
       -124.44444444, -123.33333333, -122.22222222, -121.11111111,
       -120.        , -118.88888889, -117.77777778, -116.66666667,
       -115.55555556, -114.44444444, -113.33333333, -112.22222222,
       -111.11111111, -110.        , -108.88888889, -107.77777778,
       -106.66666667, -105.55555556, -104.44444444, -103.33333333,
       -102.22222222, -101.11111111, -100.        ,  -98.88888889,
        -97.77777778,  -96.66666667,  -95.55555556,  -94.44444444,
        -93.33333333,  -92.22222222,  -91.11111111,  -90.        ,
        -88.88888889,  -87.77777778,  -86.66666667,  -85.55555556,
        -84.44444444,  -83.33333333,  -82.22222222,  -81.11111

In [117]:
def metric_to_imperial(meters):
    cms = meters * 100
    total_inches = cms / 2.54
    feet = total_inches // 12
    inches = total_inches % 12
    return feet, inches
    

In [118]:
metric_to_imperial(1.68)

(5.0, 6.141732283464563)

In [120]:
! pip install emoji # the ! is a shortcut to run the command in the terminal

twisted 18.7.0 requires PyHamcrest>=1.9.0, which is not installed.
altair 2.2.2 requires typing, which is not installed.
You are using pip version 10.0.1, however version 18.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [121]:
! python -m pip install --upgrade pip

  Using cached https://files.pythonhosted.org/packages/c2/d7/90f34cb0d83a6c5631cf71dfe64cc1054598c843a92b400e55675cc2ac37/pip-18.1-py2.py3-none-any.whl
  Found existing installation: pip 10.0.1
    Uninstalling pip-10.0.1:
      Successfully uninstalled pip-10.0.1


twisted 18.7.0 requires PyHamcrest>=1.9.0, which is not installed.
altair 2.2.2 requires typing, which is not installed.


In [122]:
import emoji

In [123]:
print(emoji.emojize(':snake: are :thumbs_up:'))

🐍 are 👍


In [124]:
%%writefile unit_conversion.py  # saves the code into *.py file

def metric_to_imperial(meters):
    cms = meters * 100
    total_inches = cms / 2.54
    feet = total_inches // 12
    inches = total_inches % 12
    return feet, inches
    

Writing unit_conversion.py


In [125]:
import unit_conversion # you can import the *.py file then as if it is a library/package

In [126]:
unit_conversion.metric_to_imperial(100)

(328.0, 1.0078740157478023)

In [128]:
import importlib                    # reloads the *.py file with the new updates in the file
importlib.reload(unit_conversion)

<module 'unit_conversion' from 'C:\\Users\\ABD01F\\Documents\\Python Charmers\\unit_conversion.py'>

In [129]:
unit_conversion.imperial_to_metric(300, 5)

91.56700000000001

In [130]:
import sys
sys.path

['',
 'C:\\Users\\ABD01F\\Documents\\Python Charmers',
 'C:\\Users\\ABD01F\\AppData\\Local\\Continuum\\anaconda3\\python37.zip',
 'C:\\Users\\ABD01F\\AppData\\Local\\Continuum\\anaconda3\\DLLs',
 'C:\\Users\\ABD01F\\AppData\\Local\\Continuum\\anaconda3\\lib',
 'C:\\Users\\ABD01F\\AppData\\Local\\Continuum\\anaconda3',
 'C:\\Users\\ABD01F\\AppData\\Local\\Continuum\\anaconda3\\lib\\site-packages',
 'C:\\Users\\ABD01F\\AppData\\Local\\Continuum\\anaconda3\\lib\\site-packages\\win32',
 'C:\\Users\\ABD01F\\AppData\\Local\\Continuum\\anaconda3\\lib\\site-packages\\win32\\lib',
 'C:\\Users\\ABD01F\\AppData\\Local\\Continuum\\anaconda3\\lib\\site-packages\\Pythonwin',
 'C:\\Users\\ABD01F\\AppData\\Local\\Continuum\\anaconda3\\lib\\site-packages\\IPython\\extensions',
 'C:\\Users\\ABD01F\\.ipython']

# Source control - avoiding the "Oh shit" moment

The idea is to snapshot at a certian point in time, so you can revert to that point at any time in the future - even if you deleted the file

common source control tool is "git" - it has 4 key commands and there are lots of GUIs so you can do it w/out the command line as well.

1. git add
2. git commit
3. git pull
4. git push

# Back to Python...

## Collections:

In [131]:
# a list is an ordered collection of data

names = []

In [132]:
front_raw = ['Greg', 'Julian', 'Seline', 'Oceane']

### Excercise: Adding data to a list

All Python objects have 'methods' - functions that work on just that object. they are accessed via dot notation


In [133]:
front_raw.append('Muhammad')

In [134]:
front_raw

['Greg', 'Julian', 'Seline', 'Oceane', 'Muhammad']

In [135]:
second_raw = ['Muhammad', 'Dave', 'Beven', 'Alex']

In [136]:
front_raw.append(second_raw)

In [137]:
front_raw

['Greg',
 'Julian',
 'Seline',
 'Oceane',
 'Muhammad',
 ['Muhammad', 'Dave', 'Beven', 'Alex']]

In [138]:
front_raw.remove(second_raw)

In [139]:
front_raw

['Greg', 'Julian', 'Seline', 'Oceane', 'Muhammad']

In [140]:
front_raw.extend(second_raw)

In [141]:
front_raw

['Greg',
 'Julian',
 'Seline',
 'Oceane',
 'Muhammad',
 'Muhammad',
 'Dave',
 'Beven',
 'Alex']

In [147]:
front_raw.append?

## Indexing - getting data out of collection

in Python are 2 ways to referance data

1.
2. Key based index 

In [148]:
front_raw[1]

'Julian'

In [149]:
len(front_raw)

9

In [150]:
del front_raw[4]

In [151]:
front_raw

['Greg', 'Julian', 'Seline', 'Oceane', 'Muhammad', 'Dave', 'Beven', 'Alex']

## Processing collections in one hit - for loop

In [153]:
for name in front_raw:
    print('Hello', name)
    

Hello Greg
Hello Julian
Hello Seline
Hello Oceane
Hello Muhammad
Hello Dave
Hello Beven
Hello Alex


In [154]:
heights = [172, 170, 180, 193, 182]

Calculate heights in feet and iches for each height

In [155]:
for h in heights:
    meters = h / 100
    print(unit_conversion.metric_to_imperial(meters))

(5.0, 7.7165354330708595)
(5.0, 6.929133858267718)
(5.0, 10.866141732283467)
(6.0, 3.9842519685039406)
(5.0, 11.653543307086608)


### Exercise: convert temperatures from C to F

In [165]:
def centigrade_to_Fehrenheit(degrees):
    return (degrees * (9 / 5)) +32

In [182]:
temprature_C = [21, 27, 20, 21, 20, 23, 22, 32]

In [187]:
results = []
for t in temprature_C:
    F = centigrade_to_Fehrenheit(t)
    results.append(F)

In [188]:
results

[69.80000000000001, 80.6, 68.0, 69.80000000000001, 68.0, 73.4, 71.6, 89.6]

In [189]:
import glob                      # dealing with lots of files instead of small lists
import os

file_size = []
for path in glob.glob('Data/*.csv'):
    file_size.append(os.path.getsize(path))

In [190]:
file_size

[211744,
 20524,
 84921,
 13735,
 48168,
 7046,
 172049,
 30973,
 100059,
 6009,
 99777,
 36071,
 13735,
 3168,
 3356,
 4384]

# Web requests - Key collections
### (page 111 in the book)

In [191]:
url = 'https://www.google.com'

In [192]:
import requests

In [193]:
requests.get(url)

<Response [200]>

In [194]:
response = requests.get(url)

In [195]:
response.headers

{'Date': 'Mon, 15 Oct 2018 05:29:00 GMT', 'Expires': '-1', 'Cache-Control': 'private, max-age=0', 'Content-Type': 'text/html; charset=ISO-8859-1', 'P3P': 'CP="This is not a P3P policy! See g.co/p3phelp for more info."', 'Content-Encoding': 'gzip', 'Server': 'gws', 'X-XSS-Protection': '1; mode=block', 'X-Frame-Options': 'SAMEORIGIN', 'Set-Cookie': '1P_JAR=2018-10-15-05; expires=Wed, 14-Nov-2018 05:29:00 GMT; path=/; domain=.google.com, NID=141=ht-qtHUgBQjF5zPVKSlkhK6Q6R_HXFtIb8VPGOIeZP97uerkTpOjGrRa8pc-haiZR1VUW6keeJUo28zkXXWGgVGueQn5VZ3TBD8VvtstQAXtwHBo0BCinBpbmgw-8G76; expires=Tue, 16-Apr-2019 05:29:00 GMT; path=/; domain=.google.com; HttpOnly', 'Alt-Svc': 'quic=":443"; ma=2592000; v="44,43,39,35"', 'Transfer-Encoding': 'chunked'}

A Python dictionary is a map of keys to values. A dictionary looks like:
    
    {Key: value, Key: value}
    
Keys are unique, values are not.    

In [196]:
response.headers['content-type']

'text/html; charset=ISO-8859-1'

## APIs

In [ ]:
%load http://bit.ly/pythoncharmersurls  # the % does not actually run the code. it just brings the information to the notebook for you to check first.
                                        # when you run this cell, the next cell will appear.
                                        # run the next cell again to confirm the URL and API key

In [199]:
# %load http://bit.ly/pythoncharmersurls
OpenWeatherMapApiKey="d1741b0c4ca70aeb629424a1ddcf28a1"
url="https://api.openweathermap.org/data/2.5/weather?q={city_name},{country_code}"
postgresqlTrainingDatabase="gis.ceebzpnip62q.ap-southeast-2.rds.amazonaws.com:5432/training"


In [201]:
response = requests.get(url)

In [202]:
response.headers['content-type']

'application/json; charset=utf-8'

In [203]:
data = response.json()  # turns text that looks like a dictionary into a dictionary

{'cod': 401,
 'message': 'Invalid API key. Please see http://openweathermap.org/faq#error401 for more info.'}

### The long and wrong way to fix the error (code: 401)

In [204]:
url

'https://api.openweathermap.org/data/2.5/weather?q={city_name},{country_code}'

In [205]:
url = 'https://api.openweathermap.org/data/2.5/weather?q=Melbourne,AU&appid=' + OpenWeatherMapApiKey

In [206]:
url

'https://api.openweathermap.org/data/2.5/weather?q=Melbourne,AU&appid=d1741b0c4ca70aeb629424a1ddcf28a1'

In [207]:
requests.get(url)

<Response [200]>

In [208]:
response = requests.get(url)

In [209]:
response.json()

{'coord': {'lon': 144.96, 'lat': -37.81},
 'weather': [{'id': 800,
   'main': 'Clear',
   'description': 'clear sky',
   'icon': '01d'}],
 'base': 'stations',
 'main': {'temp': 299.5,
  'pressure': 1013,
  'humidity': 31,
  'temp_min': 299.15,
  'temp_max': 300.15},
 'visibility': 10000,
 'wind': {'speed': 8.7, 'deg': 20, 'gust': 14.4},
 'clouds': {'all': 0},
 'dt': 1539581520,
 'sys': {'type': 1,
  'id': 8201,
  'message': 0.007,
  'country': 'AU',
  'sunrise': 1539545717,
  'sunset': 1539592637},
 'id': 2158177,
 'name': 'Melbourne',
 'cod': 200}

#### Exercise:  What temperature is in the city where you were born?

In [214]:
url = 'https://api.openweathermap.org/data/2.5/weather?q=Alexandria,EG&appid=' + OpenWeatherMapApiKey

In [215]:
response = requests.get(url)
weather = response.json()
weather['main']['temp']-273.15 ## to convert it to celsius

18.450000000000045